### PR TITLE
Cython refactor of Python C wrapping layer

### PR DESCRIPTION
**Pull request by soltanmm**
Supercedes #2134.

Begins Cython refactoring, adds Python 3 testing to the harness, and provides the fuzzy pink floating platforms upon which robotic unicorns attack.

### CODE REVIEW

File: src/python/requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 enum34==1.0.4
 futures==2.2.0
 protobuf==3.0.0a3
+cython
```
Comment (nathanielmanistaatgoogle) [inline src/python/requirements.txt]: Why not pin the version, as we do with our other dependencies?

File: src/python/src/grpc/_cython/_cygrpc/call.pyx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    cdef records.OperationTag operation_tag = records.OperationTag(tag)
+    operation_tag.operation_call = self
+    operation_tag.batch_operations = cy_operations
+    cpython.Py_INCREF(operation_tag)
+    return grpc.grpc_call_start_batch(
+        self.c_call, cy_operations.c_ops, cy_operations.c_nops,
+        <cpython.PyObject *>operation_tag)
+
+  def cancel(self,
+             grpc.grpc_status_code error_code = grpc.GRPC_STATUS__DO_NOT_USE,
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/_cygrpc/call.pyx]: Drop the whitespace on either side of the equals (here and on the next line).

File: src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+cimport cpython
+
+from grpc._cython._cygrpc cimport call
+from grpc._cython._cygrpc cimport records
+
+import threading
+import time
+
+
+cdef class _CompletionQueuePollFuture:
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx]: I strongly suspect that this is going to be a test-only structure. Do you agree? If so, can we move it to a separate test module?

File: src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      self.poll_condition.notify()
+
+    if event.type == grpc.GRPC_QUEUE_TIMEOUT:
+      return records.Event(event.type, False, None, None, None, None, None)
+    elif event.type == grpc.GRPC_QUEUE_SHUTDOWN:
+      self.is_shutdown = True
+      return records.Event(event.type, True, None, None, None, None, None)
+    else:
+      if event.tag != NULL:
+        # TODO(atash): Does this work? Do I need to <object> *then*
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx]: Can we answer this question now? If not, can we state the question more clearly? What does "object" in angle brackets mean when used as a verb?

File: src/python/src/grpc/_cython/_cygrpc/credentials.pyx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    self.c_credentials = NULL
+    self.c_ssl_pem_key_cert_pair.private_key = NULL
+    self.c_ssl_pem_key_cert_pair.certificate_chain = NULL
+    self.references = []
+
+  @property
+  def is_valid(self):
+    return self.c_credentials != NULL
+
+  @staticmethod
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/_cygrpc/credentials.pyx]: I think all the static methods on both classes in this module should instead be module-scope functions. Classes shouldn't be used as namespaces; only place behaviors inside classes when they must be placed there in order to function as they need to function.

File: src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx (hunk 4 above)
Comment (soltanmm) [inline src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx]: It's leftover cruft from before I'd written the tests. Angle brackets are used to denote casting operators in Cython the same way parentheses are used in C.

File: src/python/src/grpc/_cython/adapter_low_test.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    METHOD = 'twinkies'
+    HOST = 'hostess'
+    server_request_tag = object()
+    request_call_result = self.server.request_call(self.server_completion_queue, server_request_tag)
+
+    self.assertEquals(_types.CallError.OK, request_call_result)
+
+    client_call_tag = object()
+    client_call = self.client_channel.create_call(self.client_completion_queue, METHOD, HOST, DEADLINE)
+    client_initial_metadata = [(CLIENT_METADATA_ASCII_KEY, CLIENT_METADATA_ASCII_VALUE), (CLIENT_METADATA_BIN_KEY, CLIENT_METADATA_BIN_VALUE)]
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/adapter_low_test.py]: Does this file need to be line-wrapped to a reasonable width?

File: src/python/src/grpc/_cython/cygrpc_test.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+    request_event = self.server_completion_queue.poll(cygrpc_deadline)
+    self.assertEquals(cygrpc.CompletionType.operation_complete,
+                      request_event.type)
+    self.assertIsInstance(request_event.operation_call, cygrpc.Call)
+    self.assertIs(server_request_tag, request_event.tag)
+    self.assertEquals(0, len(request_event.batch_operations))
+    self.assertEquals(dict(client_initial_metadata),
+                      dict(request_event.request_metadata))
+    self.assertEquals(METHOD, request_event.request_call_details.method)
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_cython/cygrpc_test.py]: Throughout all this code: always leave the optional ess off the end of "assertEqual".

File: tools/run_tests/run_tests.py [hunk 8]
```diff
@@ -191,19 +191,25 @@ def __init__(self):
       self._tests = json.load(f)
 
   def test_specs(self, config, travis):
-    modules = [config.job_spec(['tools/run_tests/run_python.sh', '-m',
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: How much of these build-related configuration changes can be split out into a different changelist? Can we change the building and tooling code first and then add the code in a later change which makes a much small (something like "adding a single value to an array") tool change?

Comment (soltanmm) [inline tools/run_tests/run_tests.py]: I don't believe that the tooling changes are particularly large and don't think we'd be gaining much from separating them out into more commits.

**Comment by nathanielmanistaatgoogle** [comment]
I've kicked the failing Python build a few times but the error smells legit. What do you make of

`tools/run_tests/run_python.sh: line 39: python3.4_virtual_environment/bin/activate: No such file or directory
FAILED: grpc._cython.cygrpc_test [ret=1, pid=9638]`

in the output?

**Comment by soltanmm** [comment]
I forgot to install Python 3.4 and its `virtualenv` in the Travis yaml file.

**Comment by soltanmm** [comment]
Given the Python3.4 related changes, cc #282.

**Comment by soltanmm** [comment]
I've given up specifically using Python 3.4 (as opposed to Python 3.x) due to an incredible amount of trouble with getting Travis to provide the proper header files for building the extension without setting the entire project's language to Python. Instead this just tests against whatever version of Python 3 is available by default from `python3-dev`.

**Comment by soltanmm** [comment]
Aaannd OAuth2Client requires Python >= 3.3. What do I do with you, Travis...?

**Comment by soltanmm** [comment]
Changed to conditionally run Python `x.y` tests depending on the availability of interpreter `x.y`. Spits out a warning if a test is skipped.

**Comment by soltanmm** [comment]
(n.b. test failures are in other languages)

**Comment by jcanizales** [comment]
Only some nits. LGTM.

Note to self: adding comments on a specific commit leads to horrible UI.

**Comment by jcanizales** [comment]
Also, to be clear, commit 743c10ccd5b404edc695f2db6c8693a94b85d8c1 was already reviewed by Nathaniel?

**Comment by jcanizales** [comment]
The Travis Ruby failure is #2311. Please add issues for the other 3 failures (and note in them that Travis and Jenkins produced different results).

**Comment by soltanmm** [comment]
To the best of my knowledge Nathaniel covered 743c10ccd5b404edc695f2db6c8693a94b85d8c1 (and part of the now mutilated Py3-testing-support commit - wow GitHub's UI gets confusing with things like this) and I addressed his concerns. I'll add issues for [these](https://travis-ci.org/grpc/grpc/jobs/69783580) [three](https://travis-ci.org/grpc/grpc/jobs/69783571) [guys](https://grpc-testing.appspot.com/job/gRPC_pull_requests/602/config=tsan,language=c,platform=linux/) soon™.

**Comment by jcanizales** [comment]
That's one reason why you shouldn't rebase history that's already published :) I guess if you recover the old commit number (reflog), add your changes as a new commit, and push again, the comments will reappear.

**Comment by jcanizales** [comment]
In any case, I re-reviewed the ammended commit and it LGTM.

**Comment by soltanmm** [comment]
Jenkins failure appears to be a separate issue (#2359).

**Pull request merged**